Add draft NU6.3 deployment and Version 6 Transaction Format ZIPs (#1301)
zcash/zips: zips/draft-zodl-valargroup-ironwood-txformat.md

    ZIP: XXX

>     Title: Ironwood Transaction Format

Author: ```suggestion
    Title: Version 6 Transaction Format
```

**Common Transaction Fields** ||||
`4`                      | `header`                 | `uint32`                               | Contains the `fOverwintered` flag (bit 31, always set) and `version` (bits 30 .. 0), which MUST be `6`.
`4`                      | `nVersionGroupId`        | `uint32`                               | Version group ID (nonzero); the version 6 value fixed in [^ironwood-deploy].
`4`                      | `nConsensusBranchId`     | `uint32`                               | Consensus branch ID; MUST be the NU6.3 branch ID [^ironwood-deploy].
`4`                      | `lock_time`              | `uint32`                               | Unix-epoch UTC time or block height, encoded as in Bitcoin.
`4`                      | `nExpiryHeight`          | `uint32`                               | A block height in {1 .. 499999999} after which the transaction will expire, or 0 to disable expiry. [^zip-0203]

> `8`                      | `fee`                    | `uint64`                               | The fee to be paid by this transaction, in zatoshis. **(Under discussion — see [Open Issues](#open-issues).)**

Author: This got a bunch of pushback, let's drop it.

`1`                      | `flagsOrchard`           | `byte`                                 | An 8-bit value representing a set of flags. From LSB to MSB: `enableSpendsOrchard`, `enableOutputsOrchard`, `enableCrossAddressOrchard` (new at NU6.3); the remaining bits MUST be `0`.
`8`                      | `valueBalanceOrchard`    | `int64`                                | The net value of Orchard spends minus outputs.
`32`                     | `anchorOrchard`          | `byte[32]`                             | A root of the Orchard note commitment tree at some block height in the past.
`varies`                 | `sizeProofsOrchard`      | `compactSize`                          | Length in bytes of `proofsOrchard`. Value is `2720 + 2272 * nActionsOrchard`.
`sizeProofsOrchard`      | `proofsOrchard`          | `byte[sizeProofsOrchard]`              | Encoding of aggregated zk-SNARK proofs for Orchard Actions.
`64 * nActionsOrchard`   | `vSpendAuthSigsOrchard`  | `byte[64 * nActionsOrchard]`           | Authorizing signatures for each Orchard Action.

> `64`                     | `bindingSigOrchard`      | `byte[64]`                             | An Orchard binding signature on the SIGHASH transaction hash.

Author: These fields are absent if `nActionsOrchard = 0`; the spec has symbols indicating the conditions under which particular fields are present, and so should this table.

      only if `nActionsOrchard > 0`. If `valueBalanceOrchard` is not present, it is taken
      to be `0`.
    * The proofs in `proofsOrchard` and the signatures in `vSpendAuthSigsOrchard` each
      correspond 1:1 to the elements of `vActionsOrchard`, in the same order.

* The same rules apply to the Ironwood bundle:
    * The Ironwood fields `flagsIronwood`, `valueBalanceIronwood`, `anchorIronwood`,
      `sizeProofsIronwood`, `proofsIronwood`, and `bindingSigIronwood` are present if and
      only if `nActionsIronwood > 0`. If `valueBalanceIronwood` is not present, it is
      taken to be `0`.
    * The proofs in `proofsIronwood` and the signatures in `vSpendAuthSigsIronwood` each
      correspond 1:1 to the elements of `vActionsIronwood`, in the same order.

* In `flagsOrchard` and `flagsIronwood`, bits 3 .. 7 MUST be `0`. The semantics of
  `enableSpendsOrchard` and `enableOutputsOrchard` are as in ZIP 224 [^zip-0224]. The
  `enableCrossAddressOrchard` flag is specified in [^ironwood-circuit]; before NU6.3 this
  bit was reserved as `0`.

* For coinbase transactions, `enableSpendsOrchard` MUST be `0` in both `flagsOrchard` and

>   `flagsIronwood`.

Reviewer: Clarity: We're disabling Orchard bundles (forcing them empty) inside of coinbases after the activation of NU6.3, though it's anticipated to be documented in another ZIP.

* The `anchorOrchard` field refers to the Orchard note commitment tree; the
  `anchorIronwood` field refers to the Ironwood note commitment tree. The two trees, and

>   their nullifier sets, are disjoint.

Author: This is the wrong way of saying it; they're not "disjoint", they're unrelated.

actions — non-compact | `ZTxIdOrcActNHash` | `ZTxIdIrnActNHash`
auth                  | `ZTxAuthOrchaHash` | `ZTxAuthIrnwdHash`

When a version 6 transaction has no Ironwood actions, `ironwood_digest` and `ironwood_auth_digest`
are the hashes of empty input under the bundle and auth personalizations respectively —
`BLAKE2b-256("ZTxIdIronwd_Hash", [])` and `BLAKE2b-256("ZTxAuthIrnwdHash", [])` — distinct from the
Orchard empty-bundle digests.

### Anchor commitment (version 6)

In version 5, the Orchard anchor `anchorOrchard` is part of the **effecting data**: it is committed
in `orchard_digest` (and hence the txid), while the Orchard auth digest commits only to proofs and
signatures [^zip-0244].

In **version 6**, for **both** the Orchard and Ironwood bundles, the anchor is instead part of the
**authorizing data**. Relative to ZIP 244, for version 6 transactions:

- `orchard_digest` / `ironwood_digest` omit `anchorOrchard` / `anchorIronwood`;
- `orchard_auth_digest` / `ironwood_auth_digest` additionally commit to `anchorOrchard` /
  `anchorIronwood`.

This lets the anchor be updated (re-anchored to a more recent note commitment tree root) without

> changing the transaction ID, while signatures still bind to the specific anchor used.

Reviewer 2: err, signatures don't commit to the specific anchor, the ZKP does!

The auth digest is not hashed

Author: ```suggestion
changing the transaction ID, while the proofs still bind to the specific anchor used.
```

### Block commitments

The `hashBlockCommitments` authorizing-data commitment [^zip-0244] incorporates the version 6 auth
digest — which now includes the Ironwood auth digest, and (per the change above) the anchors — with

> no further structural change.

Author: This is not an adequate spec.

The terms "Mainnet" and "Testnet" are to be interpreted as described in § 3.12
‘Mainnet and Testnet’. [^protocol-networks]

The term "Recovery Protocol" is to be interpreted as described in [^zip-2005].

The terms below are to be interpreted as follows:

Orchard protocol (or Orchard shielded protocol)
:   The shared cryptographic design that supports the Orchard pool and the Ironwood
    pool, including the Pallas and Vesta curves, the Sinsemilla hash, the Action
    circuit (as modified by [^draft-zodl-valargroup-action-circuit-update]), the
    note, note commitment, nullifier, and key constructions, and the note encryption
    (as modified by [^zip-2005]) used for these two pools.

Orchard pool (or Orchard shielded pool)
:   The value pool, with its own note commitment tree, anchor, and chain value pool
    balance, that was introduced by ZIP 224 [^zip-0224].

Ironwood pool (or Ironwood shielded pool)
:   A new value pool of the Orchard protocol, introduced by this ZIP, with its own note
    commitment tree, anchor, and chain value pool balance, distinct from the Orchard
    pool.

# Abstract

This ZIP defines version 6 of the Zcash transaction format, to be activated at NU6.3

> [^zip-0257].

Reviewer 3: incorrectly cites NU6.2 deployment reference here.

Reviewer 2: ```suggestion
This ZIP defines version 6 of the Zcash transaction format, to be activated at NU6.3
[^zip-0258].
```

NU6.3 [^draft-zodl-valargroup-deploy-nu6.3] introduces the Ironwood shielded pool, an
Orchard-protocol successor to the Orchard pool, that ensures quantum recoverability
[^zip-2005] and has supply integrity supported from the start by formal verification
efforts [^zcash-ironwood].

This requires a new transaction version that can hold an **Ironwood component**:
a second Orchard-protocol shielded component that commits to, and spends from, the
Ironwood shielded pool, rather than the Orchard pool. The Ironwood component reuses

> the Orchard Action encoding and proof system unchanged. This ZIP defines the

Reviewer 2: Revisit proof system wording in follow-up

not by itself make Zcash secure against attacks using quantum computers, but is a
necessary and substantial step toward that goal.

The version 6 transaction format also changes the authorization structure for all
supported shielded pools (Sapling, Orchard, and Ironwood) so that anchors are
considered to be authorizing data rather than effecting data. This allows a transaction
to be pre-authorized and its proofs computed independently after choosing an anchor,
and also improves opportunities for parallelism in transaction creation.

# Privacy Implications

The turnstile mechanism reveals the amounts in each transaction that cross between pools,
including the amounts migrated into the Ironwood pool. This is considered in full detail
in [^draft-zodl-valargroup-ironwood-migration].

There is no reason to believe that the soundness vulnerability described in [^zip-0257]
could have led to any compromise of key material associated with existing addresses.
However, we encourage users to rotate addresses regularly in any case, and wallet
developers to support that practice. Further implications of address usage policy for

> privacy against quantum adversaries are discussed in [^zip-2005].

Reviewer 2: ```suggestion
Independent of the soundness bug, we recommend address rotation for quantum privacy
implications discussed in [^zip-2005].
```

Changes relative to v5 transactions should be minimized.

The version 6 transaction format must be able to carry an Ironwood-pool component in
addition to transparent-pool, Sapling-pool, and Orchard-pool components similar to
those in a version 5 transaction.

The Ironwood component must reuse the Orchard Action design and component encoding, so
that implementation is shared between them to the maximum extent possible.

The transaction identifier, signature hash, and authorizing-data commitment must
commit to the Ironwood-pool component when it is present, using personalizations
distinct from those used for the Orchard-pool component.

The changes to the authorization structure must support pre-authorizing transactions
using spending key material, and later updating the anchor and proofs (for all
supported shielded pools).

Personalizations for nodes in the digest tree must be changed when what is hashed at
that node changes.

The `flags` fields in the Orchard-pool and Ironwood-pool components must support
encoding an indication of whether outputs to the Orchard pool are required to use
a protocol-level address for which the transaction creator can authorize spends.

> 

Reviewer 2: ```suggestion
The one known Orchard-supporting hardware wallet, Keystone, must continue to be able to
sign v5 tx format Orchard -> Transparent unshielding transactions, without firmware update. 

```

# Non-requirements

This ZIP does not define the Action circuit update for NU6.3 or its `disableCrossAddress`
constraint (see [^draft-zodl-valargroup-action-circuit-update]), the activation height
or consensus branch ID (see [^draft-zodl-valargroup-deploy-nu6.3]), or wallet behaviour associated with migration
to Ironwood (see [^draft-zodl-valargroup-ironwood-migration]).

The v6 transaction format need not support ZSAs, the `zip233Amount` field, the explicit
`fee` field, or per-transparent-input sighash information that appeared in the withdrawn
ZIP 230 [^zip-0230], or those features and other extensibility affordances planned for

> ZIP 248 [^zip-0248].

Reviewer 3: question, wondering what the v6 transaction format's relation to [ZIP 246](https://zips.z.cash/zip-0246) is (specifically their overlapping digest sections), and whether it warrants a reference?

Author: ZIP 246 is now Withdrawn. I think it's not relevant.

The value 6 for the transaction version number need not avoid collisions with the
withdrawn ZIP 230 or with the current draft of ZIP 248.

# Specification

## Transaction Format

A version 6 transaction is encoded as follows. Fields up to and including the Orchard
component are as in the version 5 format [^zip-0225], except as noted; the Ironwood component
is new.

| Note | Bytes                           | Name                     | Data Type                              | Description
| ---- | ------------------------------- | ------------------------ | -------------------------------------- | -----------

> | **Common Transaction Fields** |||||

Reviewer 2: ```suggestion
| **Common Transaction Fields (unchanged from v5)** |||||
```

|      | $4$                             | `header`                 | `uint32`                               | Contains the `fOverwintered` flag (bit 31, always set) and `version` (bits 30 .. 0), which MUST be `6`.
|      | $4$                             | `nVersionGroupId`        | `uint32`                               | Version group ID (nonzero); the version 6 value fixed in [^draft-zodl-valargroup-deploy-nu6.3].
|      | $4$                             | `nConsensusBranchId`     | `uint32`                               | Consensus branch ID; MUST be the NU6.3 branch ID [^draft-zodl-valargroup-deploy-nu6.3].
|      | $4$                             | `lock_time`              | `uint32`                               | Unix-epoch UTC time or block height, encoded as in Bitcoin.
|      | $4$                             | `nExpiryHeight`          | `uint32`                               | A block height in {1 .. 499999999} after which the transaction will expire, or 0 to disable expiry. [^zip-0203]

> | **Transparent Transaction Fields** |||||

Reviewer 2: ```suggestion
| **Transparent Transaction Fields (unchanged from v5)** |||||
```

|      | $192\,\cdot$ `nSpendsSapling`   | `vSpendProofsSapling`    | `byte[192 * nSpendsSapling]`           | Encodings of the zk-SNARK proofs for each Sapling Spend.
|      | $64\,\cdot$ `nSpendsSapling`    | `vSpendAuthSigsSapling`  | `byte[64 * nSpendsSapling]`            | Authorizing signatures for each Sapling Spend.
|      | $192\,\cdot$ `nOutputsSapling`  | `vOutputProofsSapling`   | `byte[192 * nOutputsSapling]`          | Encodings of the zk-SNARK proofs for each Sapling Output.
| †    | $64$                            | `bindingSigSapling`      | `byte[64]`                             | A Sapling binding signature on the SIGHASH transaction hash.
| **Orchard Transaction Fields** |||||
|      | `varies`                        | `nActionsOrchard`        | `compactSize`                          | The number of Orchard Action descriptions in `vActionsOrchard`.
|      | $820\,\cdot$ `nActionsOrchard`  | `vActionsOrchard`        | `OrchardAction[nActionsOrchard]`       | A sequence of Orchard Action descriptions, encoded per § 7.5 ‘Action Description Encoding and Consensus’.

> | §    | $1$                             | `flagsOrchard`           | `byte`                                 | An 8-bit value representing a set of flags. From LSB to MSB: `enableSpends`, `enableOutputs`, `enableCrossAddress` (new at NU6.3); the remaining bits MUST be 0.

Reviewer 2: Is it worth noting enableCrossAddress = 0 ?

Fields marked in the Note column are conditionally present:

| Note | Description
| ---- | -----------
| †    | `valueBalanceSapling` and `bindingSigSapling` are present if and only if `nSpendsSapling + nOutputsSapling > 0`. If `valueBalanceSapling` is not present, it is taken to be 0.
| ‡    | `anchorSapling` is present if and only if `nSpendsSapling > 0`.
| §    | `flagsOrchard`, `valueBalanceOrchard`, `anchorOrchard`, `sizeProofsOrchard`, `proofsOrchard`, and `bindingSigOrchard` are present if and only if `nActionsOrchard > 0`. If `valueBalanceOrchard` is not present, it is taken to be 0.
| ◊    | `flagsIronwood`, `valueBalanceIronwood`, `anchorIronwood`, `sizeProofsIronwood`, `proofsIronwood`, and `bindingSigIronwood` are present if and only if `nActionsIronwood > 0`. If `valueBalanceIronwood` is not present, it is taken to be 0.

The encoding of the transparent and Sapling fields is unchanged from version 5
[^zip-0225]. Ironwood Action descriptions use the same `OrchardAction` encoding as
Orchard Action descriptions. Every Ironwood-pool output note uses the quantum-recoverable
note plaintext format (lead byte `0x03`) defined in ZIP 2005 [^zip-2005]; no Orchard-pool

> output note uses that format. This is the only note-level distinction between the two pools.

Reviewer 3: reminding that ZIP 2005 was written when "recoverable" meant Orchard pool notes, so it'll need commensurate updates to terminology. not sure what order those changes should land in relative to this ZIP.

Two `flagsOrchard` bits have been renamed relative to version 5: `enableSpendsOrchard` →
`enableSpends` and `enableOutputsOrchard` → `enableOutputs`. In version 6 these bits
—together with `enableCrossAddress`— are defined with the same meaning in both `flagsOrchard`
and `flagsIronwood`, so the previous `...Orchard` suffix would be misleading.

## Consensus Rules

* `nVersionGroupId` MUST equal the version 6 version group ID, and `nConsensusBranchId`
  MUST equal the NU6.3 consensus branch ID, both defined in [^draft-zodl-valargroup-deploy-nu6.3].

* The proofs in `proofsOrchard` and the signatures in `vSpendAuthSigsOrchard` each
  correspond 1:1 to the elements of `vActionsOrchard`, in the same order; likewise the
  proofs in `proofsIronwood` and the signatures in `vSpendAuthSigsIronwood` to the
  elements of `vActionsIronwood`. (Conditional presence of fields is given in the format
  table above.)

* In each of `flagsOrchard` and `flagsIronwood`:
    * Bits 3..7 inclusive MUST be 0.

>     * The semantics of the `enableSpends` and `enableOutputs` bits are as in ZIP 224 [^zip-0224].

Reviewer 3: semantics are normative to [ZIP 225](https://zips.z.cash/zip-0225) instead of ZIP 224?

Author: Actually it should be to section 7.1.2 of the protocol spec.

    * The `enableCrossAddress` bit is specified in [^draft-zodl-valargroup-action-circuit-update].
      Before NU6.3 this bit of `flagsOrchard` was reserved as 0 in v5 transactions.

* For coinbase transactions,
    * `flagsOrchard` MUST be 0.

>     * The `enableSpends` bit of `flagsIronwood` MUST be 0.

Reviewer 3: isn't this first bullet point vacuous? thought the coinbase rule being enforced was restricting Orchard bundles from appearing in coinbase transactions, so `flagsOrchard` isn't serialized there anyway, hence the vacuous statement.

Reviewer 2: Agreed, seems redundant with 
> non-normatively, these are that coinbase transactions must have an empty Orchard component;

* The `anchorOrchard` field refers to the Orchard note commitment tree, and the
  `anchorIronwood` field to the Ironwood note commitment tree. The Orchard and Ironwood
  pools have separate, independent note commitment trees and nullifier sets.

Version 4, version 5, and version 6 transactions are all valid from NU6.3 activation onward;
this ZIP defines only the version 6 format (the version 4 and version 5 formats are
unchanged). The NU6.3 consensus rules on Orchard actions apply regardless of transaction
version. In particular, the Orchard-protocol cross-address restriction is enforced for every
Orchard-pool Action mined from NU6.3 onward [^draft-zodl-valargroup-action-circuit-update],
so that it cannot be bypassed by using a version 5 transaction.

See [^draft-zodl-valargroup-deploy-nu6.3] for additional consensus requirements that apply
to **all** transactions from NU6.3 onward, regardless of transaction version. (Briefly and
non-normatively, these are that coinbase transactions must have an empty Orchard component;
that the `enableCrossAddress` bit of `flagsOrchard` must be 0; and that
$\mathsf{v}^{\textit{OrchardPoolBalance}}$ must be nonnegative.)

> ## Transaction Identifiers, Signature Hashing, and Block Commitments

Reviewer 3: misleading section title; mentions "signature hashing", but doesn't define a signature digest like [ZIP 244](https://zips.z.cash/zip-0244#signature-digest). it correctly defines `txid_digest_v6` and `auth_digest_v6`, but not what spend authorization / binding signatures actually sign in this context.

version 6 auth digest adds an Ironwood auth digest as the last child, after the Orchard one:

    auth_digest_v6 = BLAKE2b-256("ZTxAuthHash_" || consensusBranchId,
                                 transparent_auth_digest || sapling_auth_digest_v6 ||
                                 orchard_auth_digest_v6 || ironwood_auth_digest_v6)

### Anchor commitment (version 6)

In version 5, the Sapling anchor (encoded redundantly as `anchor` in each
`sapling_spends_noncompact_digest`) and the Orchard anchor (`anchorOrchard` in `orchard_digest`)
are part of the **effecting data**: they are committed to in `txid_digest`, while the Sapling and
Orchard auth digests commit only to proofs and signatures [^zip-0244-authorizingdatacommitment].

In version 6, for Sapling, Orchard, and Ironwood components, the anchor is instead part of the
**authorizing data**. Relative to ZIP 244, for version 6 transactions:

- `sapling_spends_noncompact_digest_v6`, `orchard_digest_v6`, and `ironwood_digest_v6` omit the anchor.
- `sapling_auth_digest_v6`, `orchard_auth_digest_v6`, and `ironwood_auth_digest_v6` additionally commit

>   to `anchorSapling`, `anchorOrchard`, or `anchorIronwood` respectively, after the existing fields.

Reviewer 3: small spec ambiguity: the way `sapling_auth_digest_v6` is defined reads to me like it unconditionally commits to `anchorSapling`. i get that the serialization table only includes `anchorSapling` when `nSpendsSapling > 0`, but it'd be clearer if the digest section carried that condition over explicitly. mostly just trying to keep the consistency local here rather than make readers reason across sections. for instance, i'd need to figure out how to handle the anchor in the case of a transparent to sapling transfer where`nOutputsSapling > 0 && nSpendsSapling == 0`.

        ├── ironwood_actions_memos_digest_v6
        ├── ironwood_actions_noncompact_digest_v6
        ├── flagsIronwood
        └── valueBalanceIronwood

The auth digest structure becomes:

    auth_digest_v6
    ├── transparent_scripts_digest
    ├── sapling_auth_digest_v6 (*)
    │   ├── vSpendProofsSapling
    │   ├── vSpendAuthSigsSapling
    │   ├── vOutputProofsSapling
    │   ├── bindingSigSapling
    │   └── anchorSapling (+)
    ├── orchard_auth_digest_v6 (*)
    │   ├── proofsOrchard
    │   ├── vSpendAuthSigsOrchard
    │   ├── bindingSigOrchard
    │   └── anchorOrchard (+)
    └── ironwood_auth_digest_v6 (+)
        ├── proofsIronwood
        ├── vSpendAuthSigsIronwood
        ├── bindingSigIronwood
        └── anchorIronwood

Note that the nodes under each of `sapling_digest_v6`, `orchard_digest_v6`, `ironwood_digest_v6`,
and their `auth_digest`s are present only if the corresponding component is non-empty.

### Block commitments

The `hashBlockCommitments` authorizing-data commitment [^zip-0244] incorporates the version 6 auth
digest — which now includes the Ironwood auth digest, and (per the change above) the anchors — with

> no further structural change.

Reviewer 3: "The `hashAuthDataRoot` component of `hashBlockCommitments` incorporates the version 6 auth digest".

[^zip-0221]: [ZIP 221: FlyClient - Consensus-Layer Changes](zip-0221.rst)

[^zip-0224]: [ZIP 224: Orchard Shielded Protocol](zip-0224.rst)

[^zip-0225]: [ZIP 225: Version 5 Transaction Format](zip-0225.rst)

[^zip-0230]: [ZIP 230: Withdrawn Version 6 Transaction Format](zip-0230.rst)

[^zip-0244]: [ZIP 244: Transaction Identifier Non-Malleability](zip-0244.rst)

[^zip-0244-txiddigest]: [ZIP 244: Transaction Identifier Non-Malleability — TxId Digest](zip-0244.rst#txid-digest)

[^zip-0244-authorizingdatacommitment]: [ZIP 244: Transaction Identifier Non-Malleability — Authorizing Data Commitment](zip-0244.rst#authorizing-data-commitment)

[^zip-0252]: [ZIP 252: Deployment of the NU5 Network Upgrade](zip-0252.rst)

[^zip-2005]: [ZIP 2005: Ironwood Quantum Recoverability](zip-2005.md)

[^draft-zodl-valargroup-action-circuit-update]: [NU6.3 Updates to the Orchard-protocol Action Circuit (draft)](draft-zodl-valargroup-action-circuit-update.md)

[^zip-0248]: ZIP 248: Extensible Transaction Format (unmerged; PR at <https://github.com/zcash/zips/pull/1156>)

[^zip-0257]: [ZIP 257: Deployment of the Orchard Temporary Vulnerability Mitigation and NU6.2 Network Upgrade](zip-0257.md)

> [^zip-2002]: [ZIP 2002: Explicit Fees](zip-2002.rst)

Reviewer 3: related to fees, can you remind me why the conventional fee logic in [ZIP 317](https://zips.z.cash/zip-0317) doesn't need to be modified in an update ZIP to account for ironwood bundles? it may still be worth an explicit note that Ironwood actions are counted the same as Orchard actions for logical action accounting.

Reviewer 2: Daira in ZIP call says it does

Reviewer 3: closed in https://github.com/zcash/zips/pull/1314.

constraint (see [^draft-zodl-valargroup-action-circuit-update]), the activation height
or consensus branch ID (see [^draft-zodl-valargroup-deploy-nu6.3]), or wallet behaviour associated with migration
to Ironwood (see [^draft-zodl-valargroup-ironwood-migration]).

The v6 transaction format need not support ZSAs, the `zip233Amount` field, the explicit
`fee` field, or per-transparent-input sighash information that appeared in the withdrawn
ZIP 230 [^zip-0230], or those features and other extensibility affordances planned for
ZIP 248 [^zip-0248].

The value 6 for the transaction version number need not avoid collisions with the
withdrawn ZIP 230 or with the current draft of ZIP 248.

# Specification

## Transaction Format

A version 6 transaction is encoded as follows. Fields up to and including the Orchard
component are as in the version 5 format [^zip-0225], except as noted; the Ironwood component
is new.

| Note | Bytes                           | Name                     | Data Type                              | Description
| ---- | ------------------------------- | ------------------------ | -------------------------------------- | -----------

> | **Common Transaction Fields** |||||

Reviewer 2: ```suggestion
| **Common Transaction Fields (unchanged from v5)** |||||
```

|      | $4$                             | `header`                 | `uint32`                               | Contains the `fOverwintered` flag (bit 31, always set) and `version` (bits 30 .. 0), which MUST be `6`.
|      | $4$                             | `nVersionGroupId`        | `uint32`                               | Version group ID (nonzero); the version 6 value fixed in [^draft-zodl-valargroup-deploy-nu6.3].
|      | $4$                             | `nConsensusBranchId`     | `uint32`                               | Consensus branch ID; MUST be the NU6.3 branch ID [^draft-zodl-valargroup-deploy-nu6.3].
|      | $4$                             | `lock_time`              | `uint32`                               | Unix-epoch UTC time or block height, encoded as in Bitcoin.
|      | $4$                             | `nExpiryHeight`          | `uint32`                               | A block height in {1 .. 499999999} after which the transaction will expire, or 0 to disable expiry. [^zip-0203]

> | **Transparent Transaction Fields** |||||

Reviewer 2: ```suggestion
| **Transparent Transaction Fields (unchanged from v5)** |||||
```

note plaintext format (lead byte `0x03`) defined in ZIP 2005 [^zip-2005]; no *Orchard-pool*
output note uses that format. This is the only note-level distinction between the two pools.

Two `flagsOrchard` bits have been renamed relative to version 5: `enableSpendsOrchard` →
`enableSpends` and `enableOutputsOrchard` → `enableOutputs`. In version 6 these bits
—together with `enableCrossAddress`— are defined with the same meaning in both `flagsOrchard`
and `flagsIronwood`, so the previous `...Orchard` suffix would be misleading.

## Consensus Rules

* `nVersionGroupId` MUST equal the version 6 version group ID, and `nConsensusBranchId`
  MUST equal the NU6.3 consensus branch ID, both defined in [^draft-zodl-valargroup-deploy-nu6.3].

* The proofs in `proofsOrchard` and the signatures in `vSpendAuthSigsOrchard` each
  correspond 1:1 to the elements of `vActionsOrchard`, in the same order; likewise the
  proofs in `proofsIronwood` and the signatures in `vSpendAuthSigsIronwood` to the
  elements of `vActionsIronwood`. (Conditional presence of fields is given in the format
  table above.)

* In each of `flagsOrchard` and `flagsIronwood`:
    * Bits 3..7 inclusive MUST be 0.

>     * The semantics of the `enableSpends` and `enableOutputs` bits are as in ZIP 224 [^zip-0224].

Reviewer 2: The words "flag" or "enableSpends" do not appear in 224, likely as 225 as a ref?

Reviewer 2: Agreed that this does not block merge, it's in the protocol spec

    * The `enableCrossAddress` bit is specified in [^draft-zodl-valargroup-action-circuit-update].
      Before NU6.3 this bit of `flagsOrchard` was reserved as 0 in v5 transactions.

* For coinbase transactions,

>     * `flagsOrchard` MUST be 0.

Reviewer 2: May be inconsistent, with zcash/orchard#504

Reviewer 2: Per Tal's comment, should likely just delete b/c we don't allow orchard bundles in coinbase

Reviewer 2: ```suggestion
```

* The `anchorOrchard` field refers to the Orchard note commitment tree, and the
  `anchorIronwood` field to the Ironwood note commitment tree. The *Orchard* and
  *Ironwood pools* have separate, independent note commitment trees and nullifier sets.

Version 4, version 5, and version 6 transactions are all valid from NU6.3 activation onward;
this ZIP defines only the version 6 format (the version 4 and version 5 formats are
unchanged). The NU6.3 consensus rules on Orchard actions apply regardless of transaction
version. In particular, the Orchard-protocol cross-address restriction is enforced for every
*Orchard-pool* Action mined from NU6.3 onward [^draft-zodl-valargroup-action-circuit-update],
so that it cannot be bypassed by using a version 5 transaction.

See [^draft-zodl-valargroup-deploy-nu6.3] for additional consensus requirements that apply
to **all** transactions from NU6.3 onward, regardless of transaction version. (Briefly and
non-normatively, these are that coinbase transactions must have an empty Orchard component;
that the `enableCrossAddress` bit of `flagsOrchard` must be 0; and that
$\mathsf{v}^{\textit{OrchardPoolBalance}}$ must be nonnegative.)

> ## Transaction Identifiers, Signature Hashing, and Block Commitments

Reviewer 2: ```suggestion
## Transaction Identifiers and Block Commitments
```

Version 6 transaction identifiers (txids), authorizing-data commitments ("auth digests"),
and signature hashes are computed as in ZIP 244 [^zip-0244], with two changes: an

> **Ironwood component digest** is added, and for version 6 the Orchard and Ironwood anchors

Reviewer 2: ```suggestion
**Ironwood component digest** is added, and for version 6 the Sapling, Orchard and Ironwood anchors
```

    │   ├── orchard_actions_memos_digest
    │   ├── orchard_actions_noncompact_digest
    │   ├── flagsOrchard
    │   └── valueBalanceOrchard
    └── ironwood_digest_v6 (+)
        ├── ironwood_actions_compact_digest_v6
        ├── ironwood_actions_memos_digest_v6
        ├── ironwood_actions_noncompact_digest_v6
        ├── flagsIronwood
        └── valueBalanceIronwood

The auth digest structure becomes:

    auth_digest_v6
    ├── transparent_scripts_digest
    ├── sapling_auth_digest_v6 (*)
    │   ├── vSpendProofsSapling
    │   ├── vSpendAuthSigsSapling
    │   ├── vOutputProofsSapling
    │   ├── bindingSigSapling
    │   └── anchorSapling (+)
    ├── orchard_auth_digest_v6 (*)
    │   ├── proofsOrchard
    │   ├── vSpendAuthSigsOrchard
    │   ├── bindingSigOrchard
    │   └── anchorOrchard (+)
    └── ironwood_auth_digest_v6 (+)
        ├── proofsIronwood
        ├── vSpendAuthSigsIronwood
        ├── bindingSigIronwood
        └── anchorIronwood

Note that the nodes under each of `sapling_digest_v6`, `orchard_digest_v6`, `ironwood_digest_v6`,
and their `auth_digest`s are present only if the corresponding component is non-empty.

### Block commitments

> The `hashBlockCommitments` authorizing-data commitment [^zip-0244] incorporates the version 6 auth

Reviewer 2: ```suggestion
The `hashAuthDataRoot` component of `hashBlockCommitments` [^zip-0244] incorporates the version 6 auth
```

digest — which now includes the Ironwood auth digest, and (per the change above) the anchors — with
no further structural change.

## Changes to ZIP 209

ZIP 209 [^zip-0209] is extended to track an Ironwood chain value pool balance and to require
it, like the other shielded pool balances, not to become negative.

[TODO take account of changes that should be (but are not currently) made in ZIP 256.

> The check for each pool is now that the chain value pool balance stays within [0, MAX_MONEY].]

Reviewer 2: ```suggestion
```

add

> The "Ironwood chain value pool balance" for a given block chain is the negation of the sum
> of all `valueBalanceIronwood` fields for transactions in the block chain. (Before NU6.3
> has activated, the Ironwood chain value pool balance is zero.)

In the Specification section, replace

> If any of the "Sprout chain value pool balance", "Sapling chain value pool balance", or
> "Orchard chain value pool balance" would become negative in the block chain created as a
> result of accepting a block, then all nodes MUST reject the block as invalid.

with

> If any of the "Sprout chain value pool balance", "Sapling chain value pool balance",
> "Orchard chain value pool balance", or "Ironwood chain value pool balance" would become
> negative in the block chain created as a result of accepting a block, then all nodes MUST
> reject the block as invalid.

## Changes to the Protocol Specification

Changes corresponding to the [ZIP 209 changes above](#changestozip209) are required in
§ 4.17 ‘Chain Value Pool Balances’ [^protocol-chainvalue] to define an Ironwood chain value
pool balance alongside those for the existing Sprout, Sapling, and *Orchard pools*. These

> mirror the changes above and are not spelled out here.

Reviewer 2: ```suggestion
```

the Action encoding, proving system, authorization, and note encryption (modulo the new
note plaintext format defined by [^zip-2005]) are inherited unchanged. The pools are
distinguished by their note commitment trees, nullifier sets, value pool
balances, and component position, not by separate circuits.

## Separate state

Giving the *Ironwood pool* its own note commitment tree, anchor, and nullifier set creates
a state boundary from the *Orchard pool*, so that their chain value pool balance can be
accounted for independently.

## `enableCrossAddress` polarity

This flag is encoded in the enabled sense (`1` = cross-address transfers enabled, the
normal case for *Ironwood-pool* actions; 0 = action outputs restricted to use the same
protocol-level address as the action's spend), with bit 2 reserved as 0 before NU6.3.
This is backward-compatible: an *Orchard-pool* spend after NU6.3 requires the restricted
state, which is bit 2 = 0 — exactly the value that signers treating bit 2 as a
reserved-zero bit already produce. The in-circuit constraint and the equivalent
internal `disableCrossAddress` instance value are discussed in

> [^draft-zodl-valargroup-action-circuit-update].

Reviewer 2: Move out of ZIP to Ironwood book?
